compiler.md

```diff
@@ -1,155 +1,483 @@
+<img alt="new v8 pipeline detailed" src="http://benediktmeurer.de/images/2017/architecture-20170301.png" width="70%">
 
-### Full Compiler
+### Detailed Phases of Frontend, Optimization and Backend Stages
 
-[slide](http://v8-io12.appspot.com/index.html#54)
+[slide](https://docs.google.com/presentation/d/1H1lLsbclvzyOF3IUR05ZUaZcqDxo7_-8f4yJoxdMooU/edit#slide=id.g18ceb14729_0_135)
 
-- generates code for any JavaScript
-- all code starts unoptimized
-- initial (quick) JIT 
-- is not great and knows (almost) nothing about types
-- needed to start executing code ASAP
-- uses Inline Caches (ICs) to refine knowledge about types at runtime
+<img alt="phases" src="http://benediktmeurer.de/images/2017/turbofan-20171213.png" width="70%">
 
-### Inline Caches
+## Advantages Over Old Pipeline
 
-[slide](http://v8-io12.appspot.com/index.html#55)
+[watch old architecture](https://youtu.be/HDuSEbLWyOY?t=8m51s) | [watch new architecture](https://youtu.be/HDuSEbLWyOY?t=9m21s)
+
+- reduces memory and startup overhead significantly
+- AST no longer source of truth that compilers need to agree on
+- AST much simpler and smaller in size
+- TurboFan uses Ignition bytecode directly to optimize (no re-parse needed)
+- bytecode is 25-50% the size of equivalent baseline machine code
+- combines cutting-edge IR (intermediate representation) with multi-layered translation +
+  optimization pipeline
+- relaxed [sea of nodes](#sea-of-nodes) approach allows more effective reordering and optimization when generating
+  CFG
+- to achieve that fluid code motion, control flow optimizations and precise numerical range
+  analysis are used
+- clearer separation between JavaScript, V8 and the target architectures allows cleaner, more
+  robust generated code and adds flexibility
+- generates better quality machine code than Crankshaft JIT
+- crossing from JS to C++ land has been minimized using techniques like CodeStubAssembler
+- as a result optimizations can be applied in more cases and are attempted more aggressively
+- for the same reason (and due to other improvements) TurboFan inlines code more aggressively,
+  leading to even more performance improvements
+
+### Smaller Performance Cliffs
+
+- for most websites the optimizing compiler isn't important and could even hurt performance
+  (speculative optimizations aren't cheap)
+- pages need to load fast and unoptimized code needs to run fast _enough_, esp. on mobile
+  devices
+- previous V8 implementations suffered from _performance cliffs_
+  - optimized code ran super fast (focus on peak performance case)
+  - baseline performance was much lower
+  - as a result one feature in your code that caused deoptimization would affect your app's
+    performance dramatically, i.e. 100x difference
+- TurboFan improves this as
+  - widens fast path to ensure that optimized code is more flexible and can accept more types
+    of arguments
+  - reduces code memory overhead by reusing code generation parts of TurboFan to build Ignition
```

Comment:
e.g. certain builtins (e.g. Array.p.slice) are called from both, Ignition and TurboFan code.